python/social_bot/tasks.py

```diff
@@ -851,3 +872,119 @@ def task_specific_observation(self, agent):
+            success_distance_thresh (float): consider success if the target is within this distance to the goal position
+            reward_shaping (bool): if false, the reward is -1/0/1, otherwise the 0 case will be replaced
+                with shapped reward.
+            reward_weight (float): the weight of the reward
+        """
+        super().__init__(
+            env=env, max_steps=max_steps, reward_weight=reward_weight)
+        assert self._agent.type == 'youbot_noplugin', "PickAndPlace Task only support youbot_noplugin for now"
+        self._palm_link = 'youbot_noplugin::gripper_palm_link'
+        self._finger_link_l = 'youbot_noplugin::gripper_finger_link_l'
+        self._finger_link_r = 'youbot_noplugin::gripper_finger_link_r'
+        self._object_random_range = object_random_range
+        self._place_to_random_range = place_to_random_range
+        self._mini_distance = mini_distance
+        self._success_distance_thresh = success_distance_thresh
+        self._reward_shaping = reward_shaping
+        self._env.insert_model_list(['wood_cube_5cm', 'goal_indicator'])
+        self._goal = self._world.get_model('goal_indicator')
+        self._object = self._world.get_model('wood_cube_5cm')
+
+    def run(self):
+        """ Start a teaching episode for this task. """
+        agent_sentence = yield
+        obj_pos = self._random_move_object(
+            target=self._object,
+            random_range=self._object_random_range,
+            center_pos=np.array([0, 0]),
+            mini_distance=self._mini_distance,
+            height=0)
+        goal_pos = self._random_move_object(
+            target=self._goal,
+            random_range=self._place_to_random_range,
+            center_pos=obj_pos[:2],
+            mini_distance=self._mini_distance,
+            height=0)
+        steps = 0
+        while steps < self._max_steps:
+            steps += 1
+            # get positions
+            obj_pos, _ = self._object.get_pose()
+            finger_l_pos, _ = self._agent.get_link_pose(self._finger_link_l)
+            finger_r_pos, _ = self._agent.get_link_pose(self._finger_link_r)
+            finger_pos = (
+                np.array(finger_l_pos) + np.array(finger_r_pos)) / 2.0
+            # get contacts
+            l_contact = self._agent.get_contacts(
+                'finger_cnta_l', 'wood_cube_5cm::link::collision')
+            r_contact = self._agent.get_contacts(
+                'finger_cnta_r', 'wood_cube_5cm::link::collision')
+            # check distance and contacts
+            dist = np.linalg.norm(np.array(obj_pos) - goal_pos)
+            palm_dist = np.linalg.norm(
+                np.array(obj_pos) - np.array(finger_pos))
+            gripping_feature_num = 1.0 * (
+                obj_pos[2] > 0.02) + 1.0 * l_contact + 1.0 * r_contact
+            # yield rewards
+            if gripping_feature_num >= 3 and dist < self._success_distance_thresh:
```

Comment:
The success condition should probably include the gripper not touching the object, and the object stay at that height for some time steps.  Otherwise, the gripper can just grab the 5cm block and touch the 10cm block on the side or top in an unstable way, and that would count as success?

Reply:
this one is just for placing to the goal indicator (without 10cm cube). I will add another staking cube task which have a more complex success condition.